garrysmod/lua/includes/modules/outline.lua

```diff
@@ -0,0 +1,150 @@
+
+OUTLINE_MODE_BOTH	= 0
+OUTLINE_MODE_NOTVISIBLE	= 1
+OUTLINE_MODE_VISIBLE	= 2
+
+module( "outline", package.seeall )
+
+local List, ListSize = {}, 0
+local RenderEnt = NULL
+
+local OutlineMatSettings = {
+	[ "$ignorez" ] = 1,
+	[ "$alphatest" ] = 1
+}
+
+local CopyMat		= Material( "pp/copy" )
+local OutlineMat	= CreateMaterial( "OutlineMat", "UnlitGeneric", OutlineMatSettings )
+local StoreTexture	= render.GetScreenEffectTexture( 0 )
+local DrawTexture	= render.GetScreenEffectTexture( 1 )
+
+local ENTS	= 1
+local COLOR	= 2
+local MODE	= 3
+
+function Add( ents, color, mode )
+
+	if ( ListSize >= 255 ) then return end				--Maximum 255 reference values
+	if ( !istable( ents ) ) then ents = { ents } end	--Support for passing Entity as first argument
+	if ( ents[ 1 ] == nil ) then return end				--Do not pass empty tables
+	
+	local t = {
+		[ ENTS ] = ents,
+		[ COLOR ] = color,
+		[ MODE ] = mode or OUTLINE_MODE_BOTH
+	}
+	table.insert( List, t )
+	
+	ListSize = ListSize + 1
+	
+end
+
+function RenderedEntity()
+
+	return RenderEnt
+	
+end
+
+local function Render()
+
+	local scene = render.GetRenderTarget()
+	render.CopyRenderTargetToTexture( StoreTexture )
+	
+	local w = ScrW()
+	local h = ScrH()
+	
+	render.Clear( 0, 0, 0, 0, true, true )
+
+	render.SetStencilEnable( true )
+		cam.IgnoreZ( true )
+		render.SuppressEngineLighting( true )
+	
+		render.SetStencilWriteMask( 0xFF )
+		render.SetStencilTestMask( 0xFF )
+		
+		render.SetStencilCompareFunction( STENCIL_ALWAYS )
+		render.SetStencilFailOperation( STENCIL_KEEP )
+		render.SetStencilZFailOperation( STENCIL_REPLACE )
+		render.SetStencilPassOperation( STENCIL_REPLACE )
+		
+		cam.Start3D()
+		
+			for i = 1, ListSize do
+				local v = List[ i ]
+				local mode = v[ MODE ]
+				local ents = v[ ENTS ]
+				
+				render.SetStencilReferenceValue( i )
+		
+				for j = 1, #ents do
+					local ent = ents[ j ]
+					
+					if ( !IsValid( ent ) ) then continue end
+					
+					if ( ( mode == OUTLINE_MODE_NOTVISIBLE && LocalPlayer():IsLineOfSightClear( ent ) ) || ( mode == OUTLINE_MODE_VISIBLE && !LocalPlayer():IsLineOfSightClear( ent ) ) ) then continue end
```

Comment:
LocalPlayer should be cached outside the loop too

Reply:
You can even cache the IsLineOfSightClear function to prevent extra __index calls. Ex.

```Lua
-- Outside of loop
local pPlayer = LocalPlayer()
local IsLineOfSightClear = pPlayer.IsLineOfSightClear

-- Inside of loop
if ( ( mode == OUTLINE_MODE_NOTVISIBLE && IsLineOfSightClear( pPlayer, ent ) ) || ( mode == OUTLINE_MODE_VISIBLE && !IsLineOfSightClear( pPlayer, ent ) ) ) then continue end
```